Shared/analysis/LineOfSightController.cpp

```diff
@@ -391,98 +403,129 @@ void LineOfSightController::selectOverlayIndex(int selectOverlayIndex)
     // GeoElementLineOfSight which will be added to the overlay.
-    for (auto* feature : resultsManager.m_results->iterator().features(&localParent))
+    const auto dynamicEntities = messagesOverlay->dynamicEntities();
+    for (auto* dynamicEntity : dynamicEntities)
     {
-      if (feature == nullptr)
+      if (dynamicEntity == nullptr)
         continue;
 
       // create a Line of sight from the feature to the current location
-      auto* lineOfSight = new GeoElementLineOfSight(feature, m_locationGeoElement, m_lineOfSightParent);
-      lineOfSight->setVisible(m_analysisVisible);
-      m_lineOfSightOverlay->analyses()->append(lineOfSight);
+      auto* lineOfSight = new GeoElementLineOfSight(dynamicEntity, m_locationGeoElement, m_lineOfSightParent);
+      setupNewLineOfSight(lineOfSight);
+    }
+  }
+}
 
-      m_visibleByConnections.append(connect(lineOfSight, &GeoElementLineOfSight::targetVisibilityChanged, this, [this]()
-      {
-        int visibleCount = 0;
-        auto* losList = m_lineOfSightOverlay->analyses();
-        const int count = losList->rowCount();
-        for (int i = 0; i < count; ++i)
-        {
-          auto* analysis = losList->at(i);
-          if (!analysis)
-            continue;
+bool LineOfSightController::resetAnalysis(size_t featureCount)
+{
+  constexpr int maxFeatures = 32; // Due to performance reasons, limit the number of features which can be used in the analysis
+  if (featureCount > maxFeatures)
+  {
+    emit toolErrorOccurred(QString("There are too many points in this layer (%1).\n Please choose another one with %2 or fewer points.")
+                           .arg(QString::number(featureCount), QString::number(maxFeatures)),
+                           QStringLiteral("For performance reasons, Line of Sight analysis is limited to a maximum number of features"));
+    emit selectOverlayFailed();
+    return false;
+  }
+
+  if (!m_locationGeoElement)
+    getLocationGeoElement();
+
+  if (!m_locationGeoElement)
+  {
+    emit toolErrorOccurred(QStringLiteral("Failed to get location"), QStringLiteral("Unable to find My Location GeoElement for GeoElementLineOfSight"));
+    return false;
+  }
 
-          auto* lineOfSight = qobject_cast<GeoElementLineOfSight*>(analysis);
-          if (!lineOfSight)
-            continue;
+  for (const auto& conn : m_visibleByConnections)
+    disconnect(conn);
 
-          // don't count Line of Sight from the current location
-          if (lineOfSight->observerGeoElement() == m_locationGeoElement)
-            continue;
+  m_visibleByConnections.clear();
+  setVisibleByCount(0);
 
-          if (lineOfSight->targetVisibility() == LineOfSightTargetVisibility::Visible)
-            visibleCount += 1;
-        }
+  // clear the QObject used as a parent for Line of Sight results
+  if (m_lineOfSightParent)
+  {
+    delete m_lineOfSightParent;
+    m_lineOfSightParent = nullptr;
```

Comment:
I assume the optimizer will remove this since it's overwritten on the next line. I think you can remove it.
```suggestion
```

Another option is to use a `std::unique_ptr<QObject>` as the parent. Then you just need to do a `m_lineOfSightParent = std::make_unique<QObject>();` and that does all this in a clearer way.